Source/WebKit/webpushd/PushClientConnection.mm

```diff
@@ -47,134 +50,164 @@
+    if (matchingAppsRef && CFArrayGetCount(matchingAppsRef.get())) {
+        LSASNRef asnRef = (LSASNRef)CFArrayGetValueAtIndex(matchingAppsRef.get(), 0);
+        auto bundleIdentifierRef = adoptCF((CFStringRef)_LSCopyApplicationInformationItem(sessionID, asnRef, kCFBundleIdentifierKey));
+        if (bundleIdentifierRef && CFStringGetLength(bundleIdentifierRef.get()))
+            return bundleIdentifierRef.get();
+    }
+#endif // PLATFORM(MAC)
+
+    return WebKit::codeSigningIdentifier(token);
 }
 
-void PushClientConnection::updateConnectionConfiguration(WebPushDaemonConnectionConfiguration&& configuration)
+static bool isValidPushPartition(String partition)
 {
-    if (configuration.hostAppAuditTokenData)
-        setHostAppAuditTokenData(*configuration.hostAppAuditTokenData);
+#if PLATFORM(IOS)
+    // The push partition is expected to be empty or match the format of a web clip identifier.
+    if (partition.isEmpty())
+        return true;
 
-    m_bundleIdentifierOverride = configuration.bundleIdentifierOverride;
-    m_pushPartitionString = configuration.pushPartitionString;
-    m_dataStoreIdentifier = configuration.dataStoreIdentifier;
-    m_useMockBundlesForTesting = configuration.useMockBundlesForTesting;
+    auto isASCIIDigitOrUpper = [](UChar character) {
+        return isASCIIDigit(character) || isASCIIUpper(character);
+    };
+    return partition.length() == 32 && partition.containsOnly<isASCIIDigitOrUpper>();
+#else
+    UNUSED_PARAM(partition);
+    return true;
+#endif
 }
 
-void PushClientConnection::setHostAppAuditTokenData(const Vector<uint8_t>& tokenData)
+RefPtr<PushClientConnection> PushClientConnection::create(xpc_connection_t connection, IPC::Decoder& initialMessageDecoder)
 {
-    audit_token_t token;
-    if (tokenData.size() != sizeof(token)) {
-        ASSERT_WITH_MESSAGE(false, "Attempt to set an audit token from incorrect number of bytes");
-        return;
+    if (initialMessageDecoder.messageName() != Messages::PushClientConnection::InitializeConnection::name()) {
+        RELEASE_LOG_ERROR(Push, "PushClientConnection::create failed: first message must be InitializeConnection");
+        return nullptr;
     }
 
-    memcpy(&token, tokenData.data(), tokenData.size());
-
-    if (hasHostAppAuditToken()) {
-        // Verify the token being set is equivalent to the last one set
-        audit_token_t& existingAuditToken = *m_hostAppAuditToken;
-        RELEASE_ASSERT(!memcmp(&existingAuditToken, &token, sizeof(token)));
-        return;
+    auto maybeConfiguration = initialMessageDecoder.decode<WebPushDaemonConnectionConfiguration>();
+    if (!maybeConfiguration) {
+        RELEASE_LOG_ERROR(Push, "PushClientConnection::create failed: could not decode InitializeConnection arguments");
+        return nullptr;
+    }
+    auto& configuration = *maybeConfiguration;
+
+#if USE(EXTENSIONKIT)
+    pid_t pid = xpc_connection_get_pid(connection);
```

Comment:
This is racy (pid could be recycled between the getter and the usage)

But we don't yet have a better alternative.

Reply:
Yeah unfortunately there is no more specific API we can use here.